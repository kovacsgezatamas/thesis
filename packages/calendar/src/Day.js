import React from 'react';
import PropTypes from 'prop-types';
import getDate from 'date-fns/getDate';
import isToday from 'date-fns/isToday';
import isWeekend from 'date-fns/isWeekend';
import isDate from 'date-fns/isDate';
import startOfDay from 'date-fns/startOfDay';
import endOfDay from 'date-fns/endOfDay';
import isWithinInterval from 'date-fns/isWithinInterval';
import isSameDay from 'date-fns/isSameDay';

import * as Styled from './Day.styled';

function getDayIntervalProps(date, interval) {
  let { start, end } = interval;
  end = end || start;

  if (!isDate(start) || !isDate(end)) {
    return {};
  }

  const extendedStart = startOfDay(start);
  const extendedEnd = endOfDay(end);

  return {
    isInInterval: isWithinInterval(date, { start: extendedStart, end: extendedEnd }),
    isIntervalStart: isSameDay(date, extendedStart),
    isIntervalEnd: isSameDay(date, extendedEnd),
  }
}

function Day({ date, selected, highlighted, ...restProps }) {
  if (!date) {
    return <Styled.Container {...restProps} />;
  }

  const dayNumber = getDate(date);
  const containerProps = {
    isToday: isToday(date),
    isWeekend: isWeekend(date),
    ...getDayIntervalProps(date, selected)
  };
  const highlightProps = getDayIntervalProps(date, highlighted);

  return (
    <Styled.Container {...containerProps} {...restProps}>
      {dayNumber}

      {highlightProps.isInInterval && (
        <Styled.Highlight
          isSelectionStart={containerProps.isIntervalStart}
          {...highlightProps}
        />
      )}
    </Styled.Container>
  );
}

Day.propTypes = {
  date: PropTypes.instanceOf(Date).isRequired,
  selected: PropTypes.shape({
    start: PropTypes.instanceOf(Date),
    end: PropTypes.instanceOf(Date),
  }),
  highlighted: PropTypes.shape({
    start: PropTypes.instanceOf(Date),
    end: PropTypes.instanceOf(Date),
  }),
};

Day.defaultProps = {
  selected: {},
  highlighted: {},
};

export default React.memo(Day);
export { Day };
