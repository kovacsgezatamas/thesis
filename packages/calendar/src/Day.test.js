import React from 'react';
import { shallow } from 'enzyme';
import { isToday } from 'date-fns';

import Day from './Day';

jest.mock('date-fns', () => ({ isToday: jest.fn() }));

jest.mock('./Day.styled', () => ({
  Container: () => () => 'Container',
  Highlight: () => () => 'Highlight',
}));

describe('<Day /> component', () => {
  function setupShallow(props = {}) {
    const date = new Date('2021-01-04T14:00:00Z');
    return shallow(<Day date={date} {...props} />);
  }

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should show the day number', () => {
    const wrapper = setupShallow({ date: new Date('2021-01-13T14:00:00Z') });
    const container = wrapper.find('Container');
    expect(container.children().text()).toEqual('13');
  });

  it('should render correctly', () => {
    expect(setupShallow()).toMatchSnapshot();
  });

  it('should render empty container if the date prop is not sent', () => {
    expect(setupShallow({
      date: null,
    })).toMatchSnapshot();
  })

  it('should render correctly for today', () => {
    isToday.mockReturnValueOnce(true);
    expect(setupShallow()).toMatchSnapshot();
  });

  it('should render correctly for weekend day', () => {
    expect(setupShallow({
      date: new Date('2021-01-17T17:00:00Z')
    })).toMatchSnapshot();
  });

  it('should render correctly if is in the highlighted interval', () => {
    expect(setupShallow({
      date: new Date('2021-01-13T14:00:00Z'),
      highlighted: {
        start: new Date('2021-01-09T14:00:00Z'),
        end: new Date('2021-01-22T14:00:00Z'),
      },
    })).toMatchSnapshot();
  });

  it('should render correctly if is at the beginning of the highlighted interval', () => {
    expect(setupShallow({
      date: new Date('2021-01-13T14:00:00Z'),
      highlighted: {
        start: new Date('2021-01-13T14:00:00Z'),
        end: new Date('2021-01-22T14:00:00Z'),
      },
    })).toMatchSnapshot();
  });

  it('should render correctly if is at the end of the highlighted interval', () => {
    expect(setupShallow({
      date: new Date('2021-01-13T14:00:00Z'),
      highlighted: {
        start: new Date('2021-01-08T14:00:00Z'),
        end: new Date('2021-01-13T14:00:00Z'),
      },
    })).toMatchSnapshot();
  });

  it('should render correctly if is in the selected interval', () => {
    expect(setupShallow({
      date: new Date('2021-01-13T14:00:00Z'),
      selected: {
        start: new Date('2021-01-09T14:00:00Z'),
        end: new Date('2021-01-22T14:00:00Z'),
      },
    })).toMatchSnapshot();
  });

  it('should render correctly if is at the beginning of the selected interval', () => {
    expect(setupShallow({
      date: new Date('2021-01-13T14:00:00Z'),
      selected: {
        start: new Date('2021-01-13T14:00:00Z'),
        end: new Date('2021-01-22T14:00:00Z'),
      },
    })).toMatchSnapshot();
  });

  it('should render correctly if is at the end of the selected interval', () => {
    expect(setupShallow({
      date: new Date('2021-01-13T14:00:00Z'),
      selected: {
        start: new Date('2021-01-08T14:00:00Z'),
        end: new Date('2021-01-13T14:00:00Z'),
      },
    })).toMatchSnapshot();
  });

  it('should render correctly if is in the selected and highlighted interval', () => {
    expect(setupShallow({
      date: new Date('2021-01-13T14:00:00Z'),
      selected: {
        start: new Date('2021-01-09T14:00:00Z'),
        end: new Date('2021-01-22T14:00:00Z'),
      },
      highlighted: {
        start: new Date('2021-01-08T14:00:00Z'),
        end: new Date('2021-01-23T14:00:00Z'),
      },
    })).toMatchSnapshot();
  });
});
