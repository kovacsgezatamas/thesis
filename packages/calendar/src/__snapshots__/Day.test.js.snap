// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Day /> component should render correctly 1`] = `
<Container
  isWeekend={false}
>
  4
</Container>
`;

exports[`<Day /> component should render correctly for today 1`] = `
<Container
  isToday={true}
  isWeekend={false}
>
  4
</Container>
`;

exports[`<Day /> component should render correctly for weekend day 1`] = `
<Container
  isWeekend={true}
>
  17
</Container>
`;

exports[`<Day /> component should render correctly if is at the beginning of the highlighted interval 1`] = `
<Container
  isWeekend={false}
>
  13
  <Highlight
    isInInterval={true}
    isIntervalEnd={false}
    isIntervalStart={true}
  />
</Container>
`;

exports[`<Day /> component should render correctly if is at the beginning of the selected interval 1`] = `
<Container
  isInInterval={true}
  isIntervalEnd={false}
  isIntervalStart={true}
  isWeekend={false}
>
  13
</Container>
`;

exports[`<Day /> component should render correctly if is at the end of the highlighted interval 1`] = `
<Container
  isWeekend={false}
>
  13
  <Highlight
    isInInterval={true}
    isIntervalEnd={true}
    isIntervalStart={false}
  />
</Container>
`;

exports[`<Day /> component should render correctly if is at the end of the selected interval 1`] = `
<Container
  isInInterval={true}
  isIntervalEnd={true}
  isIntervalStart={false}
  isWeekend={false}
>
  13
</Container>
`;

exports[`<Day /> component should render correctly if is in the highlighted interval 1`] = `
<Container
  isWeekend={false}
>
  13
  <Highlight
    isInInterval={true}
    isIntervalEnd={false}
    isIntervalStart={false}
  />
</Container>
`;

exports[`<Day /> component should render correctly if is in the selected and highlighted interval 1`] = `
<Container
  isInInterval={true}
  isIntervalEnd={false}
  isIntervalStart={false}
  isWeekend={false}
>
  13
  <Highlight
    isInInterval={true}
    isIntervalEnd={false}
    isIntervalStart={false}
    isSelectionStart={false}
  />
</Container>
`;

exports[`<Day /> component should render correctly if is in the selected interval 1`] = `
<Container
  isInInterval={true}
  isIntervalEnd={false}
  isIntervalStart={false}
  isWeekend={false}
>
  13
</Container>
`;

exports[`<Day /> component should render empty container if the date prop is not sent 1`] = `<Container />`;
